### PR TITLE
[GUVNOR-3326] Workbench persistence test

### PR DESCRIPTION
**Pull request by sutaakar**

### CODE REVIEW

File: test-cloud/test-cloud-remote/src/test/java/org/kie/cloud/integrationtests/persistence/WorkbenchPersistenceIntegrationTest.java [hunk 1]
```diff
@@ -0,0 +1,156 @@
...
+    public void tearDown() {
+        InstanceLogUtil.writeDeploymentLogs(workbenchWithKieServerScenario.getWorkbenchDeployment());
+        InstanceLogUtil.writeDeploymentLogs(workbenchWithKieServerScenario.getKieServerDeployment());
+        workbenchWithKieServerScenario.undeploy();
+        gitProvider.deleteGitRepository(workbenchWithKieServerScenario.getNamespace());
+    }
+
+    @Test
+    public void testWorkbenchControllerPersistence() {
+        workbenchClientProvider.createOrganizationalUnit(ORGANIZATION_UNIT_NAME, workbenchWithKieServerScenario.getWorkbenchDeployment().getUsername());
```
Comment (jhrcek) [inline test-cloud/test-cloud-remote/src/test/java/org/kie/cloud/integrationtests/persistence/WorkbenchPersistenceIntegrationTest.java]: Confusing naming. I'd expect WorkbenchClientProvider to provide some WorkbenchClient. But here it's being used as the client itself. How about renaming WorkbenchClientProvider as WorkbenchClient or similar (and simplify variable name as well)?

File: test-cloud/test-cloud-remote/src/test/java/org/kie/cloud/integrationtests/persistence/WorkbenchPersistenceIntegrationTest.java [hunk 2]
```diff
@@ -0,0 +1,156 @@
...
+        workbenchWithKieServerScenario.getWorkbenchDeployment().scale(1);
+        workbenchWithKieServerScenario.getWorkbenchDeployment().waitForScale();
+
+        KieServerInfo serverInfo = kieServerClientProvider.getKieServerClient().getServerInfo().getResult();
+        kieControllerClient.saveContainerSpec(serverInfo.getServerId(), serverInfo.getName(), CONTAINER_ID, CONTAINER_ALIAS, PROJECT_GROUP_ID, PROJECT_NAME, PROJECT_VERSION, KieContainerStatus.STARTED);
+
+        kieServerClientProvider.waitForContainerStart(CONTAINER_ID);
+
+        ServiceResponse<KieContainerResourceList> containersResponse = kieServerClientProvider.getKieServerClient().listContainers();
+        Assertions.assertThat(containersResponse.getType()).isEqualTo(ResponseType.SUCCESS);
```
Comment (jhrcek) [inline test-cloud/test-cloud-remote/src/test/java/org/kie/cloud/integrationtests/persistence/WorkbenchPersistenceIntegrationTest.java]: Static import `Assertions` please, this is unnecessarily verbose..

File: test-cloud/test-cloud-remote/src/test/java/org/kie/cloud/integrationtests/persistence/WorkbenchPersistenceIntegrationTest.java [hunk 3]
```diff
@@ -0,0 +1,156 @@
...
+
+        KieServerInfo serverInfo = kieServerClientProvider.getKieServerClient().getServerInfo().getResult();
+        String kieServerLocation = serverInfo.getLocation();
+
+        kieControllerClient.saveContainerSpec(serverInfo.getServerId(), serverInfo.getName(), CONTAINER_ID, CONTAINER_ALIAS, PROJECT_GROUP_ID, PROJECT_NAME, PROJECT_VERSION, KieContainerStatus.STARTED);
+
+        verifyOneServerTemplateWithContainer(kieServerLocation, CONTAINER_ID);
+
+        workbenchWithKieServerScenario.getWorkbenchDeployment().scale(0);
+        workbenchWithKieServerScenario.getWorkbenchDeployment().waitForScale();
```
Comment (jhrcek) [inline test-cloud/test-cloud-remote/src/test/java/org/kie/cloud/integrationtests/persistence/WorkbenchPersistenceIntegrationTest.java]: Shouldn't there be some assertion that scaling to 0 was successful?

File: test-cloud/test-cloud-remote/src/test/java/org/kie/cloud/integrationtests/persistence/WorkbenchPersistenceIntegrationTest.java [hunk 4]
```diff
@@ -0,0 +1,156 @@
...
+        workbenchWithKieServerScenario.getWorkbenchDeployment().scale(1);
+        workbenchWithKieServerScenario.getWorkbenchDeployment().waitForScale();
+
+        assertRepositoryAndProjectExists(REPOSITORY_NAME, PROJECT_NAME);
+
+        workbenchClientProvider.deployProject(REPOSITORY_NAME, PROJECT_NAME);
+
+        workbenchWithKieServerScenario.getWorkbenchDeployment().scale(0);
```
Comment (jhrcek) [inline test-cloud/test-cloud-remote/src/test/java/org/kie/cloud/integrationtests/persistence/WorkbenchPersistenceIntegrationTest.java]: This scaleDownAndUp code is repeated three times. I'd extract that to helper method.

File: test-cloud/test-cloud-remote/src/test/java/org/kie/cloud/integrationtests/persistence/WorkbenchPersistenceIntegrationTest.java (hunk 3 above)
Comment (sutaakar) [inline test-cloud/test-cloud-remote/src/test/java/org/kie/cloud/integrationtests/persistence/WorkbenchPersistenceIntegrationTest.java]: That is problematic. How to distinguish that Workbench is unavailable in different cloud environments? For example OpenShift will return HTTP 503, other providers may return HTTP 404 or any other error. The cloud environment check could be implemented in deployment ready() method and used in the test.

Comment (jhrcek) [inline test-cloud/test-cloud-remote/src/test/java/org/kie/cloud/integrationtests/persistence/WorkbenchPersistenceIntegrationTest.java]: no problem, I was just curious.

File: test-cloud/test-cloud-remote/src/test/java/org/kie/cloud/integrationtests/persistence/WorkbenchPersistenceIntegrationTest.java [hunk 5]
```diff
@@ -0,0 +1,138 @@
...
+    @Override
+    protected WorkbenchWithKieServerScenario createDeploymentScenario(DeploymentScenarioBuilderFactory deploymentScenarioFactory) {
+        return deploymentScenarioFactory.getWorkbenchWithKieServerScenarioBuilder().build();
+    }
+
+    @Before
+    public void setUp() {
+        workbenchClient = RestWorkbenchClient.createWorkbenchClient(deploymentScenario.getWorkbenchDeployment().getUrl().toString(),
+                deploymentScenario.getWorkbenchDeployment().getUsername(), deploymentScenario.getWorkbenchDeployment().getPassword());
+        kieControllerClient = new KieServerMgmtControllerClient(deploymentScenario.getWorkbenchDeployment().getUrl().toString() + "/rest/controller",
```
Comment (jakubschwan) [inline test-cloud/test-cloud-remote/src/test/java/org/kie/cloud/integrationtests/persistence/WorkbenchPersistenceIntegrationTest.java]: Why can create kieControllerClient by KieServerControllerClientProvider

File: test-cloud/test-cloud-remote/src/test/java/org/kie/cloud/integrationtests/persistence/WorkbenchPersistenceIntegrationTest.java [hunk 6]
```diff
@@ -0,0 +1,138 @@
...
+        assertRepositoryAndProjectExists(REPOSITORY_NAME, PROJECT_NAME);
+
+        scaleToZeroAndToOne(deploymentScenario.getWorkbenchDeployment());
+
+        assertRepositoryAndProjectExists(REPOSITORY_NAME, PROJECT_NAME);
+        workbenchClient.deployProject(REPOSITORY_NAME, PROJECT_NAME);
+
+        scaleToZeroAndToOne(deploymentScenario.getWorkbenchDeployment());
+
+        KieServicesClient kieServerClient = KieServerClientProvider.getKieServerClient(deploymentScenario.getKieServerDeployment());
```
Comment (jakubschwan) [inline test-cloud/test-cloud-remote/src/test/java/org/kie/cloud/integrationtests/persistence/WorkbenchPersistenceIntegrationTest.java]: Can be KieServicesClient configured in before method, or is there any reason why can't be?

File: test-cloud/test-cloud-remote/src/test/java/org/kie/cloud/integrationtests/persistence/WorkbenchPersistenceIntegrationTest.java (hunk 5 above)
Comment (sutaakar) [inline test-cloud/test-cloud-remote/src/test/java/org/kie/cloud/integrationtests/persistence/WorkbenchPersistenceIntegrationTest.java]: Good point, implemented.

File: test-cloud/test-cloud-remote/src/test/java/org/kie/cloud/integrationtests/persistence/WorkbenchPersistenceIntegrationTest.java (hunk 6 above)
Comment (sutaakar) [inline test-cloud/test-cloud-remote/src/test/java/org/kie/cloud/integrationtests/persistence/WorkbenchPersistenceIntegrationTest.java]: Yes, implemented.

**Pull request merged**